util/topgen/templates/plic_all_irqs_test.c.tpl

```diff
@@ -0,0 +1,246 @@
+  dif_rv_plic_irq_id_t plic_irq_id;
+  CHECK(dif_rv_plic_irq_claim(&plic, kHart, &plic_irq_id) == kDifRvPlicOk,
+        "dif_rv_plic_irq_claim failed");
+
+  // Check if it is the right peripheral.
+  top_earlgrey_plic_peripheral_t peripheral = (top_earlgrey_plic_peripheral_t)
+      top_earlgrey_plic_interrupt_for_peripheral[plic_irq_id];
+  CHECK(peripheral == peripheral_expected,
+        "Interurpt from incorrect peripheral: exp = %d, obs = %d",
+        peripheral_expected, peripheral);
+<%text>
+#define PERIPHERAL_ISR(_peripheral, _handle, _plic_start_irq)                 \
+  do {                                                                        \
+    /* Correlate the interrupt fired at PLIC with the peripheral. */          \
+    dif_##_peripheral##_irq_t irq = (dif_##_peripheral##_irq_t)(              \
+        plic_irq_id - (dif_rv_plic_irq_id_t)_plic_start_irq);                 \
+    CHECK(irq == _peripheral##_irq_expected,                                  \
+          "Incorrect " #_handle " IRQ received at PLIC: exp = %d, obs = %d",  \
+          _peripheral##_irq_expected, irq);                                   \
+    _peripheral##_irq_serviced = irq;                                         \
+                                                                              \
+    /* Check if the same interrupt fired at the peripheral as well. */        \
+    /* TODO: read the whole status register to confirm none other pending. */ \
+    bool is_pending;                                                          \
+    CHECK(dif_##_peripheral##_irq_is_pending(&_handle, irq, &is_pending) ==   \
+              kDifOk,                                                         \
+          "dif_" #_peripheral "_irq_is_pending failed");                      \
+    CHECK(is_pending, "Interrupt %d fired at PLIC did not fire at " #_handle, \
+          irq);                                                               \
+                                                                              \
+    /* Clear the interrupt at the peripheral. */                              \
+    CHECK(dif_##_peripheral##_irq_acknowledge(&_handle, irq) == kDifOk,       \
+          "dif_" #_peripheral "_irq_acknowledge failed");                     \
+  } while (0)
+</%text>\
+
+  switch (peripheral) {
+% for p in helper.irq_peripherals:
+<%
+  comment = '' if p.name in enabled_peripherals else '// '
+%>\
+    ${comment}case ${p.plic_name}:
+    ${comment}  PERIPHERAL_ISR(${p.name}, ${p.inst_name}, ${p.plic_start_irq});
+    ${comment}  break;
+% endfor
+    default:
+      LOG_FATAL("ISR is not implemented!");
+      test_status_set(kTestStatusFailed);
+  }
+
+#undef PERIPHERAL_ISR
+
+  // Complete the IRQ at PLIC.
+  CHECK(dif_rv_plic_irq_complete(&plic, kHart, plic_irq_id) == kDifRvPlicOk,
+        "dif_rv_plic_irq_complete failed");
+}
+
+/**
+ * Enables all IRQs at PLIC.
+ */
+static void plic_irqs_enable(void) {
```

Comment:
Functions like this one should probably be hoisted into another target, just like the macros. Let's keep things that don't depend on template input separate.